### PR TITLE
Introduce Comparison API v1

### PR DESCRIPTION
**Pull request by kritika-singh3**
Issue: #3624
API Doc PR: https://github.com/gocd/api.go.cd/pull/468

Description (Edited):
- has a GET call only
- format: `/api/pipelines/:pipeline_name/compare/:from_counter/:to_counter`

E.g.
Request:
```
curl 'http://localhost:8153/go/api/pipelines/downstream2/compare/1/1' \
  -u 'username:password' \
  -H 'Accept: application/vnd.go.cd+json'
```

Response:
```
{
  "_links" : {
    "self" : {
      "href" : "http://localhost:8153/go/api/pipelines/:pipeline_name/compare/:from_counter/:to_counter"
    },
    "doc" : {
      "href" : "https://api.gocd.org/19.9.0/#compare"
    }
  },
    "pipeline_name" : "downstream2",
    "from_counter" : 1,
    "to_counter" : 1,
    "is_bisect" : false,
    "changes" : [ {
      "material" : {
        "type" : "git",
        "attributes" : {
          "destination" : null,
          "filter" : null,
          "invert_filter" : false,
          "name" : null,
          "auto_update" : true,
          "url" : "http://github.com/gocd/gocd",
          "branch" : "master",
          "submodule_folder" : null,
          "shallow_clone" : false
        }
      },
      "revision" : [ {
        "revision_sha" : "c35c40097e8adb6e0ffc0d570631f57ca3841735",
        "modified_by" : "Username <abc@gmail.com>",
        "modified_at" : "2019-10-04T16:28:07Z",
        "commit_message" : "some commit msg"
      } ]
    }, {
      "material" : {
        "type" : "git",
        "attributes" : {
          "destination" : null,
          "filter" : null,
          "invert_filter" : false,
          "name" : null,
          "auto_update" : true,
          "url" : "http://github.com/gocd/docs.go.cd",
          "branch" : "master",
          "submodule_folder" : null,
          "shallow_clone" : false
        }
      },
      "revision" : [ {
        "revision_sha" : "0a98aaaec4d38473ba2d5ccd1047dc80f3f6a365",
        "modified_by" : "preview[bot] <emailid>",
        "modified_at" : "2019-09-30T05:52:42Z",
        "commit_message" : "some msg"
      } ]
    }, {
      "material" : {
        "type" : "git",
        "attributes" : {
          "destination" : null,
          "filter" : null,
          "invert_filter" : false,
          "name" : null,
          "auto_update" : true,
          "url" : "http://github.com/gocd/api.go.cd",
          "branch" : "master",
          "submodule_folder" : null,
          "shallow_clone" : false
        }
      },
      "revision" : [ {
        "revision_sha" : "2e6fa03df438fb6a5eaf7018ae77a77101734f0c",
        "modified_by" : "preview[bot] <emailid>",
        "modified_at" : "2019-10-04T10:45:52Z",
        "commit_message" : "some msg"
      } ]
    }, {
      "material" : {
        "type" : "dependency",
        "attributes" : {
          "pipeline" : "up42",
          "stage" : "up42_stage",
          "name" : "up42",
          "auto_update" : true
        }
      },
      "revision" : [ {
        "revision" : "up42/2/up42_stage/1",
        "pipeline_counter" : "2",
        "completed_at" : "2019-10-07T06:15:53Z"
      } ]
    }, {
      "material" : {
        "type" : "dependency",
        "attributes" : {
          "pipeline" : "downstream1",
          "stage" : "up42_stage",
          "name" : "downstream1",
          "auto_update" : true
        }
      },
      "revision" : [ {
        "revision" : "downstream1/1/up42_stage/1",
        "pipeline_counter" : "1",
        "completed_at" : "2019-10-07T06:16:48Z"
      } ]
    } ]
}
```

### CODE REVIEW

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+            before("/*", mimeType, this.apiAuthenticationHelper::checkUserAnd403);
+
+            get("", mimeType, this::index);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        String pipelineName = request.params("pipelineName");
+        Integer fromCounter = Integer.valueOf(request.params("fromCounter"));
+        Integer toCounter = Integer.valueOf(request.params("toCounter"));
```
Comment (GaneshSPatil) [inline api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: Any reason for picking `camelCase` over `snake_case`?

I know its just a parameter value. But as per our [API documentation](https://api.gocd.org/current), we've been using `snake_case` everywhere.

Example: look for `:pipeline_name` on https://api.gocd.org/current

Comment (kritika-singh3) [reply api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: No, there was no specific reason for this. Will change it to `snake_case`.

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+            before("", mimeType, this.apiAuthenticationHelper::checkUserAnd403);
+            before("/*", mimeType, this.apiAuthenticationHelper::checkUserAnd403);
+
+            get("", mimeType, this::index);
+        });
+    }
+
+    public String index(Request request, Response response) throws IOException {
+        String pipelineName = request.params("pipeline_name");
+        Integer fromCounter = Integer.valueOf(request.params("from_counter"));
```
Comment (maheshp) [inline api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: How do we handle invalid counters, consider throwing a UnprocessableEntityException for a invalid counter and if less than 1.

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/representers/ComparisonRepresenter.java [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+import com.thoughtworks.go.domain.MaterialRevision;
+import com.thoughtworks.go.spark.Routes;
+
+import java.util.List;
+import java.util.function.Consumer;
+
+public class ComparisonRepresenter {
+    public static void toJSON(OutputWriter outputWriter, String pipelineName, Integer fromCounter, Integer toCounter, boolean isBisect, List<MaterialRevision> materialRevisions) {
+        outputWriter.addLinks(addLinks())
+                .addChild("_embedded", embeddedWriter -> embeddedWriter.add("pipeline_name", pipelineName)
```
Comment (maheshp) [inline api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/representers/ComparisonRepresenter.java]: I don't think we need it to be `_embedded`.

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+        Integer toCounter = Integer.valueOf(request.params("to_counter"));
+        HttpLocalizedOperationResult result = new HttpLocalizedOperationResult();
+
+        if (fromCounter < 1 || toCounter < 1) {
+            throw new UnprocessableEntityException("The instance counters cannot be less than 1.");
+        }
+
+        boolean isBisect = pipelineService.isPipelineBisect(pipelineName, fromCounter, toCounter);
```
Comment (maheshp) [inline api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: We will have to document what does bisect means in this context.

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+
+    @Override
+    public void setupRoutes() {
+        path(controllerBasePath(), () -> {
+            before("", mimeType, this::setContentType);
+            before("/*", mimeType, this::setContentType);
+
+            before("", mimeType, this::verifyContentType);
+
+            before("", mimeType, this.apiAuthenticationHelper::checkUserAnd403);
```
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: The [`ChangesetService`](https://github.com/gocd/gocd/blob/master/server/src/main/java/com/thoughtworks/go/server/service/ChangesetService.java#L72) has a check for `viewPermission`. So thought, adding another check would be redundant. WDYT?

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java (hunk 2 above)
Comment (kritika-singh3) [reply api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: Done.

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java (hunk 4 above)
Comment (kritika-singh3) [reply api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: Document as in API documentation?

File: api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java (hunk 5 above)
[the remark this answers is not in the document]
Comment (kritika-singh3) [reply api/api-compare-v1/src/main/java/com/thoughtworks/go/apiv1/compare/CompareControllerV1.java]: Added the check.

**Pull request merged**